File: dart/utils/tree_node.dart
library tree_node;
import 'dart:math';
part 'btree_printer.dart';

class TreeNode {
  int data;
  TreeNode left;
  TreeNode right;
  TreeNode parent;
  int _size = 0;

  TreeNode(this.data) : _size = 1;

  void setLeftChild(TreeNode left) {
    this.left = left;
    if (left != null) {
      left.parent = this;
    }
  }

  void setRightChild(TreeNode right) {
    this.right = right;
    if (right != null) {
      right.parent = this;
    }
  }

  void insertInOrder(int d) {
    if (d <= data) {
      if (left == null) {
        setLeftChild(new TreeNode(d));
      } else {
        left.insertInOrder(d);
      }
    } else {
      if (right == null) {
        setRightChild(new TreeNode(d));
      } else {
        right.insertInOrder(d);
      }
    }

    _size++;
  }

  int size() => _size;

  bool isBST() {
    if (left != null) {
      if (data < left.data || !left.isBST()) {
        return false;
      }
    }

    if (right != null) {
      if (data >= right.data || !right.isBST()) {
        return false;
      }
    }

    return true;
  }

  int height() {
    int leftHeight = left != null ? left.height() : 0;
    int rightHeight = right != null ? right.height() : 0;
    return 1 + max(leftHeight, rightHeight);
  }

  TreeNode find(int d) {
    if (d == data) {
      return this;
    } else if (d <= data) {
      return left != null ? left.find(d) : null;
    } else if (d > data) {
      return right != null ? right.find(d) : null;
    }

    return null;
  }

  static TreeNode createMinimalBST(List<int> arr, {int start: null, int end: null}) {
    if (start == null || end == null) {
      return createMinimalBST(arr, start: 0, end: arr.length -1);
    }

    if (end < start) {
      return null;
    }

    int mid = (start + end) ~/ 2;
    TreeNode n = new TreeNode(arr[mid]);
    n.setLeftChild(createMinimalBST(arr, start: start, end: mid - 1));
    n.setRightChild(createMinimalBST(arr, start: mid + 1, end: end));
    return n;
  }


  void print() {
    BTreePrinter.printNode(this);
  }

}

File: c-sharp/ctci.Library/TreeNode.cs
using System;
using System.Diagnostics;

namespace ctci.Library
{
    [DebuggerDisplay("{Data}")]
    public class TreeNode
    {
        public int Data {get; set;}
        public TreeNode Left {get; set;}
	    public TreeNode Right {get; set;}
	    public TreeNode Parent {get; set;}
	    public int Size {get; set;}

	    public TreeNode(int data) 
        {
		    Data = data;
		    Size = 1;
	    }
	
	    public void SetLeftChild(TreeNode left) 
        {
		    Left = left;

		    if (left != null) 
            {
			    left.Parent = this;
		    }
	    }
	
	    public void SetRightChild(TreeNode right) 
        {
		    Right = right;
            
		    if (right != null) 
            {
			    right.Parent = this;
		    }
	    }
	
	    public void InsertInOrder(int data) 
        {
		    if (data <= Data) 
            {
			    if (Left == null) 
                {
				    SetLeftChild(new TreeNode(data));
			    } 
                else 
                {
				    Left.InsertInOrder(data);
			    }
		    } 
            else 
            {
			    if (Right == null) 
                {
				    SetRightChild(new TreeNode(data));
			    }
                else 
                {
				    Right.InsertInOrder(data);
			    }
		    }

		    Size++;
	    }
	
	    public bool IsBst() 
        {
		    if (Left != null)
            {
			    if (Data < Left.Data || !Left.IsBst()) 
                {
				    return false;
			    }
		    }
		
		    if (Right != null) 
            {
			    if (Data >= Right.Data || !Right.IsBst()) 
                {
				    return false;
			    }
		    }		
		
		    return true;
	    }
	
	    public int Height() 
        {
		    var leftHeight = Left != null ? Left.Height() : 0;
		    var rightHeight = Right != null ? Right.Height() : 0;

		    return 1 + Math.Max(leftHeight, rightHeight);
	    }
	
	    public TreeNode Find(int data) 
        {
		    if (data == Data) 
            {
			    return this;
		    } 
            else if (data <= Data) 
            {
			    return Left != null ? Left.Find(data) : null;
		    } 
            else if (data > Data) 
            {
			    return Right != null ? Right.Find(data) : null;
		    }

		    return null;
	    }
	
	    private static TreeNode CreateMinimalBst(int[] array, int start, int end)
        {
		    if (end < start) 
            {
			    return null;
		    }

		    var mid = (start + end) / 2;
		    var treeNode = new TreeNode(array[mid]);
		    treeNode.SetLeftChild(CreateMinimalBst(array, start, mid - 1));
		    treeNode.SetRightChild(CreateMinimalBst(array, mid + 1, end));

		    return treeNode;
	    }
	
	    public static TreeNode CreateMinimalBst(int[] array) 
        {
		    return CreateMinimalBst(array, 0, array.Length - 1);
	    }
	
	    public void Print() 
        {
            BTreePrinter.PrintNode(this);
	    }
    }
}